### PR TITLE
sorter/leveldb(ticdc): separate read from table sorter

### LINKED ISSUE #4631: Separate read and write path for leveldb sorter
Separate read and write path to improve read and write throughput.

We plan to add two actors reader and writer, and let table sorter be a thin shim, which delegates AddEntry to writer and Output to reader.

| | |
| - | - |
|image| image |

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

Issue Number: close #4631 

### What is changed and how it works?

* Sorter delegates AddEntry to writer, Output to reader.
* Sorter system spawns read system for readers.
* Reader receive ReadTs message and try to output
  resolved events to output channel.

After this PR, leveldb sorter will be refactor into following 

| | |
| - | - |
|image| image |

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

1. drop database test
2. create database test
3. create a changefeed
4. load and run tpcc workload
5. compare upstream and downstream

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

### Release note 

```release-note
Imporve thoughput of leveldb sorter by separating write and read path.
```

### CODE REVIEW

File: cdc/sorter/sorter.go [hunk 1]
```diff
@@ -30,7 +30,8 @@ type EventSorter interface {
 	// Returns error if the sorter is closed or context is done
 	TryAddEntry(ctx context.Context, entry *model.PolymorphicEvent) (bool, error)
 	// Output sorted events, orderd by commit ts.
-	// It may output a dummy event, a zero resolved ts event, to detect whether
-	// output is available.
+	//
+	// Callers must not caching the returned channel, as sorter may not output
+	// any resolved events if callers skip calling `Output`.
```
Comment (sdojjy) [inline cdc/sorter/sorter.go:35]: Can you add a code example about how to use this function in the comment?

Comment (sdojjy) [inline cdc/sorter/sorter.go:35]: Can you add a code example about how to use this function in the comment?

File: cdc/sorter/leveldb/table_sorter.go [hunk 2]
```diff
@@ -15,307 +15,90 @@ package leveldb
...
-	case c.errCh <- err:
-	default:
-		// It means there is an error already.
-	}
-}
-
-// Sorter accepts out-of-order raw kv entries and output sorted entries
-type Sorter struct {
+// reader accepts out-of-order raw kv entries and output sorted entries
+type reader struct {
```
Comment (sdojjy) [inline cdc/sorter/leveldb/table_sorter.go:33]: there is a writer.go,  maybe rename the file to reader.go is better

File: cdc/sorter/leveldb/sorter.go [hunk 3]
```diff
@@ -0,0 +1,262 @@
+// Copyright 2021 PingCAP, Inc.
```
Comment (sdojjy) [inline cdc/sorter/leveldb/sorter.go:1]:
```suggestion
// Copyright 2022 PingCAP, Inc.
```

File: cdc/sorter/leveldb/system/system.go [hunk 4]
```diff
@@ -227,6 +235,10 @@ func (s *System) Stop() error {
 	if err != nil {
 		return errors.Trace(err)
 	}
+	err = s.ReaderSystem.Stop()
```
Comment (sdojjy) [inline cdc/sorter/leveldb/system/system.go:238]: shall we ignore the error and stop all actor systems

File: cdc/sorter/leveldb/sorter.go [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+	return true, nil
+}
+
+// Output returns the sorted raw kv output channel
+func (ls *Sorter) Output() <-chan *model.PolymorphicEvent {
+	// Notify reader to read sorted events
+	msg := actormsg.SorterMessage(message.Task{
+		UID:     ls.uid,
+		TableID: ls.tableID,
+		ReadTs:  message.ReadTs{},
```
Comment (sdojjy) [inline cdc/sorter/leveldb/sorter.go:239]: so, with empty `ReadTs` works like a tick or notification message

File: cdc/sorter/leveldb/sorter.go [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+	})
+	// It's ok to ignore error, as reader is either channel full or stopped.
+	// If it's channel full, it has been notified by others, and caller will
+	// receive new resolved events eventually.
+	_ = ls.readerRouter.Send(ls.readerActorID, msg)
```
Comment (sdojjy) [inline cdc/sorter/leveldb/sorter.go:244]: we call `_ = w.readerRouter.Send(w.readerActorID, msg)` in writer too,  do we need some improvement to reduce the actor messages?

File: cdc/sorter/leveldb/table_sorter.go (hunk 2 above)
Comment (overvenus) [reply cdc/sorter/leveldb/table_sorter.go:33]: I'm going to do in the next PR. By doing so, the diff is huge in this PR, hard to review.

File: cdc/sorter/leveldb/system/system.go (hunk 4 above)
Comment (overvenus) [reply cdc/sorter/leveldb/system/system.go:238]: It may leaks some resouces, like iterators.

File: cdc/sorter/leveldb/sorter.go (hunk 5 above)
Comment (overvenus) [reply cdc/sorter/leveldb/sorter.go:239]: Exactly!

File: cdc/sorter/leveldb/sorter.go (hunk 6 above)
Comment (overvenus) [reply cdc/sorter/leveldb/sorter.go:244]: Good point! I'll add a TODO here.

**Comment by zhaoxinyu** [comment]
A unit test `TestMySQLSinkFlushResovledTs` failed.

**Comment by sdojjy** [comment]
> A unit test `TestMySQLSinkFlushResovledTs` failed.
> 
> [2022-03-01T15:55:53.225Z] Error Trace: mysql_test.go:1180 [2022-03-01T15:55:53.225Z] Error: Not equal: [2022-03-01T15:55:53.225Z] expected: 0x0 [2022-03-01T15:55:53.225Z] actual : 0x1 [2022-03-01T15:55:53.225Z] Test: TestMySQLSinkFlushResovledTs

I will fix this ut in  https://github.com/pingcap/tiflow/issues/4748

**Review by sdojjy — APPROVED** [review]

**Pull request merged**